File: packages/dart/lib/src/project.dart
class Project {
  // ===== 标识 =====
  final String id;
  final String name;
  final String title;
  final String description;

  // ===== 审计 =====
  final String? createdBy;
  final DateTime? createdAt;
  final String? updatedBy;
  final DateTime? updatedAt;

  const Project({
    required this.id,
    required this.name,
    required this.title,
    this.description = '',
    this.createdBy,
    this.createdAt,
    this.updatedBy,
    this.updatedAt,
  });

  factory Project.fromJson(Map<String, dynamic> json) {
    DateTime? parseDateTime(String key) {
      final v = json[key];
      if (v == null) return null;
      if (v is DateTime) return v;
      if (v is String) return DateTime.tryParse(v);
      return null;
    }

    return Project(
      id: json['id'] as String,
      name: json['name'] as String,
      title: json['title'] as String,
      description: json['description'] as String? ?? '',
      createdBy: json['createdBy'] as String?,
      createdAt: parseDateTime('createdAt'),
      updatedBy: json['updatedBy'] as String?,
      updatedAt: parseDateTime('updatedAt'),
    );
  }

  Map<String, dynamic> toJson() {
    String? formatDateTime(DateTime? dt) =>
        dt?.toIso8601String();

    final map = <String, dynamic>{
      'id': id,
      'name': name,
      'title': title,
      'description': description,
    };
    if (createdBy != null) map['createdBy'] = createdBy;
    if (createdAt != null) map['createdAt'] = formatDateTime(createdAt);
    if (updatedBy != null) map['updatedBy'] = updatedBy;
    if (updatedAt != null) map['updatedAt'] = formatDateTime(updatedAt);
    return map;
  }
}

File: packages/python/src/quanttide_project/models/project.py
from datetime import datetime
from typing import Any, Optional

from pydantic import BaseModel, ConfigDict, field_serializer


class Project(BaseModel):
    model_config = ConfigDict(frozen=True, populate_by_name=True)

    id: str
    name: str
    title: str
    description: str = ""

    created_by: Optional[str] = None
    created_at: Optional[datetime] = None
    updated_by: Optional[str] = None
    updated_at: Optional[datetime] = None

    @field_serializer("created_at", "updated_at")
    def serialize_datetime(self, v: Optional[datetime]) -> Optional[str]:
        if v is None:
            return None
        return v.isoformat()

    def to_dict(self) -> dict[str, Any]:
        d = self.model_dump(mode="json", exclude_none=True)
        for key in ("created_at", "updated_at"):
            if key in d and d[key] is None:
                del d[key]
        return d
